File: ines.cpp
#include <array>
#include <algorithm>
#include <memory>
#include <cstdio>
#include <cstring>
#include <cstdint>

#include "junknes.h"
#include "ines.hpp"
#include "util.hpp"

using namespace std;

namespace{
    constexpr char INES_MAGIC[4] = { 'N', 'E', 'S', '\x1A' };

    union INesFlags{
        uint8_t raw;
        explicit INesFlags(uint8_t value) : raw(value) {}
        BitField8<0>   mirror_v;
        BitField8<1>   sram;
        BitField8<2>   trainer;
        BitField8<3>   mirror_four;
        BitField8<4,4> mapper;
    };
}

bool ines_split(const char* path,
                std::array<std::uint8_t, 0x8000>& prg,
                std::array<std::uint8_t, 0x2000>& chr,
                JunknesMirroring& mirror)
{
    unique_ptr<FILE, decltype(&fclose)> in(fopen(path, "rb"), fclose);
    if(!in) return false;

    uint8_t hdr[16];
    if(fread(hdr, 1, sizeof(hdr), in.get()) != sizeof(hdr)) return false;
    if(memcmp(hdr, INES_MAGIC, 4) != 0) return false;

    unsigned int prg_count = hdr[4];
    unsigned int chr_count = hdr[5];
    INesFlags flags(hdr[6]);
    // ゴミが書かれている場合があるのでByte7は見ない

    if(!(prg_count == 1 || prg_count == 2)) return false;
    if(!(chr_count == 0 || chr_count == 1)) return false; // 一部のテストROMはCHRを持たない
    if(flags.sram) return false;
    if(flags.trainer) return false;
    if(flags.mirror_four) return false;
    if(flags.mapper != 0) return false;

    if(fread(prg.data(), 1, 0x4000*prg_count, in.get()) != 0x4000*prg_count) return false;
    if(chr_count)
        if(fread(chr.data(), 1, 0x2000, in.get()) != 0x2000) return false;

    // 16K PRGはミラーして32Kにする
    if(prg_count == 1)
        copy_n(prg.begin(), 0x4000, prg.begin()+0x4000);
    // CHRがなければ0で埋める
    if(chr_count == 0)
        chr.fill(0);

    mirror = flags.mirror_v ? JUNKNES_MIRROR_V : JUNKNES_MIRROR_H;

    return true;
}

File: ines.py
# -*- coding: utf-8 -*-

#---------------------------------------------------------------------
# 基本的なINES仕様のみ対応
# マッパー0のみ対応
#---------------------------------------------------------------------

import junknes

class Error(Exception): pass

def _read_exact(in_, n_bytes):
    buf = in_.read(n_bytes)
    if len(buf) != n_bytes: raise Error("incomplete file")
    return buf

def ines_split(in_):
    INES_MAGIC = b"NES\x1A"

    header = _read_exact(in_, 16)
    if header[:4] != INES_MAGIC: raise Error("INES magic not found")

    prg_count = header[4]
    chr_count = header[5]
    flags     = header[6]
    # ゴミが書かれている場合があるのでByte7は見ない

    if not (1 <= prg_count <= 2): raise Error("invalid prg count: {}".format(prg_count))
    # 一部のテストROMはCHRを持たない
    if not (0 <= chr_count <= 1): raise Error("invalid chr count: {}".format(chr_count))
    if flags & 2: raise Error("sram is not supported")
    if flags & 4: raise Error("trainer is not supported")
    if flags & 8: raise Error("fourscreen is not supported")
    mapper = flags >> 4
    if mapper != 0: raise Error("only mapper 0 is supported")

    # 16K PRGはミラーして32Kに
    prg  = _read_exact(in_, 0x4000*prg_count)
    if prg_count == 1: prg *= 2
    # CHRがなければ0フィル
    chr_ = _read_exact(in_, 0x2000*chr_count) if chr_count else b"\0"*0x2000

    mirror = junknes.JUNKNES_MIRROR_V if (flags&1) else junknes.JUNKNES_MIRROR_H

    return prg, chr_, mirror
